Update simd-json to make serenity compile again

kangalio opened the pull request:
And fix an unsoundness (lol)

I found out that serenity doesn't compile with the simd_json feature in poise https://github.com/serenity-rs/poise/actions/runs/6158123854/job/16710276804

kangalio committed:
Update simd-json to make serenity compile again

And fix an unsoundness (lol)

kangalio commented:
After reading https://github.com/serenity-rs/serenity/issues/2474#issuecomment-1637242395, I realize serenity actually re-exports the raw serialization and deserialization functions from the underlying JSON library in serenity::json::prelude. As such, any upgrade to serde-json or simd-json is a breaking change >:(

We could hack around this unfortunate decision by replacing the re-exported simd_json::from_str with a custom impl that keeps the old signature (at the cost of a soundness hole, that probably no one ever will encounter because this is such a niche serenity API). We should also maybe deprecate the re-exports from the underlying JSON libraries?

mkrasnitski commented:
Is changing the trait bound to `DeserializeOwned` a breaking change? I'm not sure - since we're dealing with JSON, are there any types that don't implement DeserializeOwned?

kangalio commented:
serenity::json::from_str isn't public API! Only serenity::json::prelude::from_str is. The two differ in that the public one is directly re-exported from serde_json or simd_json, whereas the private one is a wrapper function around serde_json or simd_json

---

I have another hack idea to fix the semver issue: depend on both simd_json 0.4 and 0.10 - use 0.10 internally, re-export 0.4. It's silly but it's a silly solution to a silly problem (namely, serenity publicly re-exporting implementation details)

kangalio committed:
Re-export old simd-json to avoid breaking change

arqunis commented on src/json.rs:
```diff
@@ -145,7 +146,7 @@ pub mod prelude {
         to_vec_pretty,
     };
     #[cfg(feature = "simd-json")]
```
I believe the feature flag name here should be the one with the underscore:
```suggestion
    #[cfg(feature = "simd_json")]
```

Otherwise, only `simd-json` gets pulled in.

kangalio replied:
Ah good catch. That's actually fine though (it was in the code before this PR, even) - the `simd-json` feature flag is an automatic feature from Cargo that cannot be turned off that's always set if simd-json was pulled in. The `simd_json` is our own feature flag that aliases to `simd-json` for naming consistency. So gating by `simd-json` or `simd_json` does the same..

It is confusing though, I give you that.

Should I change all places in the codebase that gate by `simd-json` to say `simd_json`?

arqunis replied:
> Should I change all places in the codebase that gate by `simd-json` to say `simd_json`?

We should be using our feature flag, i.e. `simd_json`. So yes, please. Not just in code, but also [in the test CI job](https://github.com/serenity-rs/serenity/blob/next/.github/workflows/ci.yml#L43), which is why it currently fails.

kangalio replied:
Done

mkrasnitski commented on src/json.rs:
```diff
@@ -62,19 +60,22 @@ where
+    T: DeserializeOwned,
 {
-    Ok(simd_json::from_str(s)?)
+    // Have to clone the argument because simd_json mutates the argument
+    // And we can't just take &mut str either because simd_json mutates the string such that it may
+    // not be UTF-8 afterwards, which is unsound
+    Ok(simd_json::from_slice(&mut s.as_bytes().to_vec())?)
```
Does `from_slice` work with multi-byte characters? I think that passing in a `&mut s.to_string()` would still be fine here, since even if the string becomes invalid UTF-8, it gets dropped immediately afterward.

kangalio replied:
> Does from_slice work with multi-byte characters?

Yes:
![Screenshot_20230919_033925](https://github.com/serenity-rs/serenity/assets/21220820/3a7f498c-e9af-4d73-abf8-a63b8e31aa02)

> I think that passing in a &mut s.to_string() would still be fine here, since even if the string becomes invalid UTF-8, it gets dropped immediately afterward.

I originally thought the mere existence of a non UTF-8 str is instant UB but I looked it up and it seems you're right

Still, that would need an unsafe block. Using `from_slice` needs no unsafe block and has no performance penalty (I expect?

kangalio committed:
Global simd-json -> simd_json replace

As per arqunis' / acdenis' request.

Also add a feature gated compile_error! to make sure people don't accidentally enable simd-json instead of simd_json again

kangalio committed:
Oops

mkrasnitski commented:
Ok, turns out this approach doesn't entirely work because having things like `OwnedValue` and `Object` imported from simd-json 0.10 is not compatible with methods like `to_string` which are imported from 0.4. What would need to happen is for two separate preludes to be created, one which is `pub` that re-exports everything from 0.4, and one which is internal containing everything from 0.10. Additionally, the pub types like `Value`, `JsonMap`, `JsonError`, etc. would need to be set to their versions from `simd_json_old`, which means that their new versions should therefore also be exported only in the "internal-only" prelude.

Considering that the `json` macro is broken on old simd-json versions, I'm really not sure if it's worth fixing this just to re-export a broken API, because the later rebases of `next` on top of that work  would be pretty painful. Maybe we just bite the bullet and re-export the new versions of the functions, except for `from_str` for which the wrapper is already `pub` anyways? I don't think any of the other function signatures have been broken by simd-json.

The other fixes in this PR (like the feature flag typo fix) are fine and can be left in.

mkrasnitski commented:
Any updates on this?

kangalio commented:
I'm currently not motivated to implement that separate prelude thing, and pushing breaking changes into `current` feels wrong to me. Maybe it's ok if I just close this PR, let simd-json continue to be broken on current, and wait for `next` instead ™ ?

mkrasnitski commented:
If you'd like, we can close this and split it into separate PRs.

kangalio commented:
Split into what PRs? What would they do?

mkrasnitski commented:
Replacing the uses of the incorrect `simd-json` feature with `simd_json` is relatively simple and worth incorporating.

kangalio commented:
More trouble than its worth tbh. That replacing is a "pick up a piece of litter in your way" level of change
